Print Channel Name as String in Gossip Log

lindluni opened the pull request:
Channel names are currently byte arrays in the logs, convert them to string.

Signed-off-by: Brett Logan <brett.t.logan@ibm.com>

kchristidis commented:
@jeffgarratt @jyellick @srderson: you guys are good candidates for reviewing this. Please let me know of any issues you see with the code and I'll gladly fix them. Thanks!

PS: @srderson as I noted on Slack, I expect the CI test to fail because I have moved the `discovery` directory, which seems to be hardcoded in in the CI's check scripts. With the modified scripts, this PR passes all tests (at least on my end).

binhn commented:
LGTM - Thanks.

jyellick commented on core/discovery/discovery.go:
```diff
@@ -0,0 +1,109 @@
+	if _, ok := di.nodes[address]; ok {
+		di.nodes[address] = false
+	}
+	return !di.nodes[address]
```
Is there a scenario where this does not return true?

kchristidis replied:
Fixed in https://github.com/hyperledger/fabric/pull/2016/commits/9aae07b9d6fd76f2ee45d7c2635846772474ad92

jyellick commented on core/discovery/discovery.go:
```diff
@@ -0,0 +1,109 @@
+		for {
+			pick = di.seq[di.random.Intn(len(di.nodes))]
+			if di.nodes[pick] {
+				break
```
It looks like the same node could be picked multiple times, is that okay here?

kchristidis replied:
Good catch, thank you. Fixed in https://github.com/hyperledger/fabric/pull/2016/commits/fc0d06e79d20f3d37a4d7d5a52b68354d9dc06d3

jyellick commented on core/peer/peer.go:
```diff
@@ -322,6 +322,14 @@ func (p *PeerImpl) GetPeers() (*pb.PeersMessage, error) {
+func getPeerAddresses(peersMsg *pb.PeersMessage) []string {
+	var addresses []string
+	for _, v := range peersMsg.GetPeers() {
+		addresses = append(addresses, v.Address)
```
Not certain how frequently this is called, but the `addresses` length looks like it could be computed to reduce allocations rather than appended up to size.

kchristidis replied:
Fixed in https://github.com/hyperledger/fabric/pull/2016/commits/91aa90b9f4fbe3452510398f1745b5127367bebb

jyellick commented on core/peer/peer.go:
```diff
@@ -808,3 +794,109 @@ func (p *PeerImpl) GetTransactionResultByUUID(txUuid string) (*pb.TransactionRes
+	p.discHelper = discovery.NewDiscoveryImpl()
+	p.discPersist = viper.GetBool("peer.discovery.persist")
+	if !p.discPersist {
+		peerLogger.Warning("Discovery list will not be persisted to disk")
```
Should this also abort loading the list from disk?

kchristidis replied:
I claim that it should not because you may run the peer with persistence and then decide for whatever reason to switch it off at a later point. Let me know if you disagree. Also note that at some point we'll need to develop a workflow that removes nodes from the discovery list. (The method is already there, but the process that will result in this invocation has not been defined yet.)

jyellick replied:
Works for me, just wanted to make sure it had been considered.

jyellick commented:
Looks good to me, thanks for the updates @kchristidis

yacovm commented:
Maybe just [add](https://github.com/hyperledger/fabric/blob/master/gossip/common/common.go#L46) a Stringer?

Commit:
Add stringer for channelID type

Channel names are currently byte arrays in the logs,
add stringer to convert the to string when printing

Signed-off-by: Brett Logan <brett.t.logan@ibm.com>

lindluni commented:
🤦  Done

yacovm commented:
timer.Reset()

lindluni commented:
@Mergifyio backport release-2.2
